### PR TITLE
Upgrade to PT 2.12

### PR DESCRIPTION
**Pull request by ani300**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

This PR upgrades torch-spyre to use Pytorch 2.12. I also used the excuse to do a dependency pass and update our other dependencies. If CI catches something we can decide if we want to avoid some of them.

### CODE REVIEW

**Comment by ani300** [comment]
@mcalman what do we need for profiler if we are finally on 2.12? I know the new API for PrivateUse1 becomes available.

File: pyproject.toml [hunk 1]
```diff
@@ -89,17 +89,17 @@ disable_error_code = ["attr-defined"]
 
 [tool.pytest.ini_options]
 filterwarnings = [
-    # PyTorch 2.10.0+: Suppress torch.jit.script_method DeprecationWarning triggered by
+    # PyTorch 2.12.0+: Suppress torch.jit.script_method DeprecationWarning triggered by
```
Comment (yoheiueda) [inline pyproject.toml]:
```suggestion
    # Suppress torch.jit.script_method DeprecationWarning triggered by
```

This comment causes false positives, so we can remove it to simplify future upgrades.

File: pyproject.toml [hunk 2]
```diff
@@ -89,17 +89,17 @@ disable_error_code = ["attr-defined"]
...
     # importing torch._inductor.compile_fx (which transitively imports torch.utils.mkldnn).
-    # https://github.com/pytorch/pytorch/blob/v2.11.0/torch/jit/_script.py#L366
+    # https://github.com/pytorch/pytorch/blob/v2.12.0/torch/jit/_script.py#L366
```
Comment (yoheiueda) [inline pyproject.toml]:
```suggestion
    # https://github.com/pytorch/pytorch/blob/70d99e9/torch/jit/_script.py#L366
```

We can use a permalink here.

**Comment by HarikrishnanBalagopal** [comment]
PTAL https://github.com/torch-spyre/torch-spyre/pull/2274

**Comment by SilverSoldier** [comment]
> @mcalman what do we need for profiler if we are finally on 2.12? I know the new API for PrivateUse1 becomes available.

@ani300 Just tested your PR + our changes in #1856 for enabling profiler with 2.12 and it works smoothly. 

Thanks!

File: pyproject.toml (hunk 2 above)
Comment (ani300) [reply pyproject.toml]: the current link is a tag too, not a branch head, so it should be as permanent

Comment (yoheiueda) [reply pyproject.toml]: That’s correct.

My point is that we could avoid updating the string `v2.11.0` in the URL when upgrading PyTorch. Version tags in links may be confusing for AI, as they can appear to need updating to the latest version. Changing the tag may also cause the referenced lines to shift between releases.

Using a permalink could help keep the reference stable, so the link continues to point to the same location and avoids confusing the AI.

**Comment by tdoublep** [comment]
Just FYI upstream vLLM has **not yet** moved to PT 2.12 (draft PR: https://github.com/vllm-project/vllm/pull/42848). Will these changes ensure backwards compatibility with PT 2.11? I am worried that merging this PR could prevent spyre-inference from pulling in any latest changes from torch-spyre side.

**Comment by ani300** [comment]
@tdoublep they in fact break backwards compatibility... I can refactor to use both 2.11 and 2.12 APIs everywhere it needs to, but there are a lot of breaking changes for 2.12 that affect us: decompositions are handled through a completely different path, profiling is also handled differently, symbolic shapes hints have a new API as well... these are all changes we (as IBM) have requested from pytorch to better integrate out-of-tree devices.

@bohnstingl can you think of a way to do decompositions that works in both 2.11 and 2.12?
@SilverSoldier is your PR going to be compatible with both 2.12 and 2.11?

**Comment by bohnstingl** [comment]
> can you think of a way to do decompositions that works in both 2.11 and 2.12?

@ani300, yes the decomposition behavior that we do in torch-spyre in 2.12 is not BC, but we don't have to use the feature from upstream. I believe there is a way to still use the old-style for the decompositions, as upstream itself is BC in that regards. However, from the spyre-inference perspective we don't touch the decompositions at the moment and this change should be transparent to us. 

I more echo the worry from @tdoublep that vLLM upstream appears to have not yet moved to PT2.12. I quickly tested this out and compiled vllm as well as our spyre-inference plugin myself with PT2.12 and it appears to be working, see https://github.com/torch-spyre/spyre-inference/issues/214#issuecomment-4612261480

**Comment by bohnstingl** [comment]
Just FYI, my upstream correction to the decomposition handling landed in PyTorch, see https://github.com/pytorch/pytorch/pull/185909.
Unfortunately, we it will take another PyTorch release to become effective. The planned release date is 8/7/26, see https://dev-discuss.pytorch.org/t/pytorch-release-2-13-key-dates/3390

**Comment by SilverSoldier** [comment]
> @tdoublep they in fact break backwards compatibility... I can refactor to use both 2.11 and 2.12 APIs everywhere it needs to, but there are a lot of breaking changes for 2.12 that affect us: decompositions are handled through a completely different path, profiling is also handled differently, symbolic shapes hints have a new API as well... these are all changes we (as IBM) have requested from pytorch to better integrate out-of-tree devices.
> 
> @bohnstingl can you think of a way to do decompositions that works in both 2.11 and 2.12? @SilverSoldier is your PR going to be compatible with both 2.12 and 2.11?

The change introducing the Profiler Registration macro is only part of the 2.12 release. So no, unfortunately my changes will not work with 2.11 (i.e. can't enable profiler when compiling, will fail)

**Comment by bohnstingl** [comment]
@ani300, PyTorch is planning a patch release, `PyTorch 2.12.1`, with the GA on June 17th. I try to get my decomposition PR cherry-picked, see https://github.com/pytorch/pytorch/issues/186805#issuecomment-4672532422

**Review by dgrove-oss — COMMENTED** [review]
Initial coding agent review.

Overview

This PR upgrades torch-spyre from PyTorch 2.11 to 2.12. It handles several breaking API changes in Inductor and fixes resulting regressions. In addition to the version bump, it updates all dependent tooling (mypy 2.1, pre-commit 4.6, uv 0.11, types-PyYAML, etc.).

---
Key Changes

1. size_hint → optimization_hint API rename — PT 2.12 renamed sizevars.size_hint() to optimization_hint() throughout Inductor. Correctly updated in superdsc.py, pass_utils.py, views.py, and spyre_kernel.py. The fix in pass_utils.py:concretize_index is particularly good — it also adds a guard to exclude indirect index symbols (gather/scatter tmp0 variables) which PT 2.12 would now erroneously concretize rather than raising, silently breaking named-dim propagation for gathers.

2. Decomposition architecture refactor (decompositions.py, patches.py, __init__.py) — The old approach mutated PyTorch's global decomposition registry inside a context manager, which caused the FX graph cache to bypass itself (BypassFxGraphCache("Failed to pickle cache key")). The new approach:
- Builds a fresh, independent decomposition table via get_spyre_decomp_table(), never mutating the global registry
- Threads it into Inductor via a new _spyre_inner_compile wrapper that sets get_decomp_fn to a picklable module-level callable
- Replaces the _OPWrapper per-instance compilation with lazy compilation (first call compiles, subsequent calls reuse)
- Renames register_spyre_decomposition → register_spyre_decompositions (plural, consistent with the table name)

This is a significant correctness fix and architectural improvement.

3. addmm regression — PT 2.12 re-fuses add(input, mm(a, b)) into aten.addmm.default in the post-grad pattern matcher, defeating the upstream decomposition. Two coordinated fixes:
- spyre_decompositions_to_exclude now drops aten.mm and aten.bmm (prevents PT 2.12's K==1 mm/bmm decomposition that breaks stick expressions)
- New decompose_addmm FX pass in temp_passes.py undoes the re-fusion before Spyre's lowerings run
- convert_constant_with_graph_node is now in temp_passes.py (was it previously defined elsewhere? This looks new)

4. SDPA output layout fix — PT 2.12 changed the meta kernel for _scaled_dot_product_fused_attention_overrideable to use alloc_with_matching_layout(query, ...), so the output must now be BHSE-contiguous. The old code did transpose(1,2).clone().transpose(1,2) unnecessarily. The fix (returning out directly from matmul) is correct and simpler.

5. Fallback lowerings for ops with in-tree decompositions — PT 2.12 added core-aten decompositions for several ops that Spyre falls back to (arange, tril, triu, isin, etc.). The new register_fallback_over_decomp() in lowering.py pre-registers these with override_decomp=True before Inductor's asserting auto-make_fallback path can trigger.

6. SpyreAsyncCompile now inherits AsyncCompile and delegates wait() to super(). Previously wait() was a no-op pass, meaning async-compiled kernels were never waited on. This is a silent correctness fix.

7. max_dim_int64_fallback / min_dim_int64_fallback — These custom op bodies previously raise RuntimeError(...). PT 2.12 routes non-Spyre tensor inputs (e.g. compare-with-cpu test paths) through CompositeExplicitAutograd, so they must now return a real result. The fix is correct.

8. to_dtype lowering — PT 2.12 adds a use_compute_types kwarg; the Spyre lowering and kernel handler now accept and forward it.

9. spyre_to Dynamo tracing — torch._dynamo.allow_in_graph(torch.Tensor.to) is added so Dynamo traces .to as a leaf rather than inlining the Python wrapper, preventing graph breaks during D2D dtype casts.

10. Tolerance loosening in tensor layout tests — atol=1e-3, rtol=1e-3 replaces the default atol=1e-5 for fp16 round-trip tests. The comment explains it well.

---
Issues

BLOCKER

- _OPWrapper device check uses != instead of is not (decompositions.py:1122-1130): The check getattr(x.device, "type", None) != DEVICE_NAME will always be True for None (when x has no .device.type), but the logic flow is fine since the outer isinstance(x, torch.Tensor) guard precedes it. However the devs list comprehension on the error path iterates leaves twice (unnecessary): once to build leaves and once for the error list. Not a blocker, minor.
- convert_constant_with_graph_node appears to be entirely new code added to temp_passes.py but is already referenced in passes.py — it must have been there before (I can't confirm without git history). If it truly is new, there are no tests for it beyond the decompose_addmm integration path.

SUGGESTIONS

- psutil pin removed (pyproject.toml:88): psutil~=7.2.1 is changed to just psutil (unpinned). While the lockfile pins it at 7.2.2, this loosens the constraint
in pyproject.toml in a way that could cause drift for users instaonsider psutil>=7.0 or keeping the tighter constraint.
- expect_fail per-op string matching (utils_inductor.py:483-487): The new op_case = f"{op_name}_{test_case}" matching is a good enhancement. However, the logic
checks op_case in expect_fail twice (once for the condition, onceplication but readable.
- test_tensor_layout.py tolerance change (atol=1e-3, rtol=1e-3): For a round-trip fp16 copy test, atol=1e-3 is quite loose — fp16 machine epsilon is ~9.77e-4,
so this is barely above epsilon. The comment attributes this to "st 1e-5" due to fp16 storage granularity (~1/2048 = ~4.9e-4). The1e-3 bound is plausible but worth a second look — was this validated empirically or chosen conservatively? If it's empirical, adding a comment with the observed
max delta would help reviewers in the future.
- torch._dynamo.allow_in_graph(torch.Tensor.to) (_monkey_patch.py): This marks the global torch.Tensor.to as allowed in graph for all users of the process, not
just Spyre. The comment acknowledges this is necessary, but it wo description since it's a global side-effect that could affectother backends sharing the process.
- uv.lock resolution marker change: The lock now splits on pythondition to sys_platform. This is a uv internal change but worthconfirming intentional (was uv 0.11.15 the actual cause, or was the Python 3.15 marker added manually?).

QUESTIONS

- The addmm_decomp function in decompositions.py (lines ~290-323 in the old code) was removed. Is the new decompose_addmm FX pass in temp_passes.py a complete
replacement, or are there eager-mode paths where addmm can still The old version handled out= kwarg; the new FX pass does notappear to handle aten.addmm.out. Is that overload excluded from the fallback path?
- The copy_from_d2d change in eager.py uses torch._C._dispatch_tlython") to detect no_dispatch(). Is this the recommended upstreamdetection pattern, or is there a more stable API? This looks like a private symbol.
- The test_exp_sin 3D shape is now expect_fail: ["3d"]. The commeent past tolerance." Was the root cause investigated — is this aPT 2.12 numerics change in the CPU reference (changing the baseline the test compares against) or an actual Spyre kernel regression?

---
Verdict
                                                                                                                                                               APPROVE with suggestions. This is a well-executed version upgradeure refactor is a genuine correctness improvement beyond justcompatibility. The per-file comments explaining each PT 2.12 behavioral change are thorough and will help future maintainers. Address the psutil pin loosening and confirm the addmm.out overload coverage before merging.

File: torch_spyre/_inductor/temp_passes.py [hunk 3]
```diff
@@ -363,3 +363,164 @@ def _unflatten_bmm_batch_dims(
...
+                    replacement.meta["val"] = torch.empty_like(out_meta, device="meta")
+                copy_fx_custom_meta(node, replacement)
+
+        node.replace_all_uses_with(replacement)
+        graph.erase_node(node)
+
+    graph.lint()
+
+
+def convert_constant_with_graph_node(graph: torch.fx.Graph) -> None:
```
Comment (dgrove-oss) [inline torch_spyre/_inductor/temp_passes.py]: I think `convert_constant_with_graph_node` is getting pulled back in via a bad merge?  I believe we pushed this down to the looplevelir portion of the compiler and dropped it from the FX graph layer are in temp_passes.

**Comment by ani300** [comment]
**Re: `addmm.out` overload coverage (question)**

`decompose_addmm` intentionally handles only `aten.addmm.default` — that's the overload PT 2.12's post-grad pattern matcher re-fuses `add(input, mm(a, b))` into. The `.out` overload doesn't go through that re-fusion path, so it doesn't need undoing. Confirmed empirically: `test_addmm_out` passes in the (green) `Test Inductor Ops Matmul` CI job on 2.12. Eager-mode `addmm` also dispatches through the normal PrivateUse1 path, not this FX pass, so there's no eager gap.

**Comment by ani300** [comment]
**Re: `copy_from_d2d` no_dispatch detection uses a private symbol (question)**

`torch._C._dispatch_tls_is_dispatch_key_excluded("Python")` is private, agreed. It's the pragmatic way to detect that we're inside a `no_dispatch()` region (which excludes the Python dispatch key) from within an op kernel — there is no public predicate for "am I under no_dispatch". The alternative (attempt `copy_from_d2d` and catch the re-entrancy failure) is worse. Context: `copy_from_d2d` requires `torch.compile`, which can't run inside `no_dispatch()` (e.g. FakeTensorMode constant propagation), so we fall back to a CPU roundtrip there. Added a comment flagging the private-API dependency so it's revisited if a stable API appears.

**Comment by ani300** [comment]
**Re: `test_fallback` 3d xfail — was the root cause investigated? (question)**

It's a PT 2.12 **CPU-reference** numerics change, not a Spyre kernel regression. This is one of three pre-existing PT-2.12 fp16 edge cases documented and xfailed in commit `3a2d482` ("Xfail three pre-existing PT 2.12 numerical edge cases"): the 3D fp16 `(8, 16, 256)` `exp → sin (CPU fallback) → exp` case drifts a single element (~0.34 abs, 1/32768 elements) past the 0.1 tolerance; 1D/2D pass. PT 2.12 changed the reference decomposition/promotion, so the CPU baseline the test compares against moved — the Spyre result itself is unchanged. Chose xfail over loosening tolerance because it's an isolated single-element edge case, matching how the sibling cases were handled. Added the `3a2d482` reference to the inline comment.

**Comment by ani300** [comment]
**Re: `test_tensor_layout.py` tolerance `atol=1e-3` (suggestion)**

Chosen empirically. On-device fp16 storage rounds to a `~1/2048 ≈ 4.9e-4` granularity, and the observed max round-trip delta on these `torch.rand([512, 256])`-scale tensors was ~4.9e-4 (the failures were near-zero values where the default `atol=1e-5` couldn't absorb the drift). `atol=1e-3` sits just above that without being loose enough to mask a real regression. Updated the comment to record the measured ~4.9e-4 delta so it's clear this is empirical, not arbitrary.

**Comment by ani300** [comment]
**Re: `torch._dynamo.allow_in_graph(torch.Tensor.to)` is a global side-effect (suggestion)**

Correct that it's process-global. It's intended/acceptable: torch-spyre already monkey-patches `torch.Tensor.to = spyre_to` globally at import, so this backend already owns `.to`'s behavior in-process. `allow_in_graph` only changes how Dynamo traces it (as a leaf) rather than breaking on the C++ `orig_to`; for cpu/other-backend tensors `spyre_to` falls through to `orig_to` semantics and leaf-tracing `.to` is harmless (Dynamo's default treatment anyway). Expanded the comment to note the global scope and rationale.

**Comment by ani300** [comment]
**Re: `_OPWrapper` device check (`!=` vs `is not`) + double leaf iteration (minor)**

Kept the `!=`: it compares the device-*type* string against the `DEVICE_NAME` constant, so value equality (`!=`) is correct — `is not` would be wrong for a string comparison. Added a comment noting this is deliberate. The `isinstance(x, torch.Tensor)` guard short-circuits before the `getattr(..., None)`, so the `None` case never reaches the comparison. (`leaves` is already built once; the error-path list comprehension only runs on the raise path.)

**Comment by ani300** [comment]
**Re: `psutil` pin loosening (suggestion)**

Intentional — `psutil` is left unpinned to match the upstream PyTorch shared dependency constraint (same treatment as the other deps torch shares). Keeping it unpinned avoids conflicting with torch's own resolution on upgrade; the lockfile still pins the concrete version. Leaving as-is.

**Review by dgrove-oss — COMMENTED** [review]
One more question about temp_passes that was still pending; not sure why this didn't post last night.

File: torch_spyre/_inductor/temp_passes.py (hunk 3 above)
Comment (ani300) [reply torch_spyre/_inductor/temp_passes.py]: done, removed

**Comment by tdoublep** [comment]
We have merged a [PR](https://github.com/torch-spyre/spyre-inference/pull/357) in spyre-inference that removes our dependency on precompiled CPU wheels of upstream vLLM. This gives us more flexibility in terms of which torch version we can use, and in particular means we can bump the torch version before upstream vLLM does. Nothing blocking this anymore from our side.

File: torch_spyre/_monkey_patch.py [hunk 4]
```diff
@@ -209,6 +209,22 @@ def spyre_empty(
     torch.Tensor.device_tensor_layout = device_tensor_layout
     torch.Tensor._spyre_tensor_patched = True
     torch.Tensor.to = spyre_to
+    # Dynamo cannot trace INTO the Python ``spyre_to``: it inlines the wrapper,
```
Comment (thoangtrvn) [inline torch_spyre/_monkey_patch.py:212]: Is this an issue only for PT 2.12? or it's a bug that has not been addressed in PT 2.11?

**Review by thoangtrvn — DISMISSED** [review]
LGTM. Awesome!

File: torch_spyre/_monkey_patch.py (hunk 4 above)
Comment (ani300) [reply torch_spyre/_monkey_patch.py:212]: this is specific to PT 2.12. It doesn't happen in 2.11, probably due to some change in dynamo between versions.

**Comment by dgrove-oss** [comment]
Code review reports that:

torch_spyre/_inductor/codegen/superdsc.py:602 — _concretize_for_sdsc was changed from V.graph.sizevars.size_hint(expr) to V.graph.sizevars.optimization_hint(expr). This function's own docstring calls it "the final concretization point in the pipeline" whose output must be a fully concrete integer for the downstream DeepTools/SDSC compiler. But per PT 2.12's own docs, optimization_hint is explicitly "intended for optimization purposes only," and unlike size_hint/guarding_hint_or_throw it will silently return a heuristic or fallback value (even sys.maxsize for int_oo, or a fallback for NaN) for unbacked symbols rather than resolving to the true value or raising. That's the wrong tradeoff at a correctness-critical boundary — it can now silently emit a bogus iteration-space size into SDSC JSON instead of failing loudly. Notably, size_hint is untouched and still used identically for the same "concretize for codegen" purpose in work_division.py, spyre_kernel.py, pass_utils.py, and views.py — size_hint was not removed/deprecated in 2.12, so this wasn't a forced migration. This looks like an unintentional edit that should be reverted to size_hint.

**Review by mudhakar — DISMISSED** [review]
LGTM - nothing to add beyond @dgrove-oss 's review, which @ani300 has already addressed ^^

**Review by dgrove-oss — DISMISSED** [review]
Approving.  We don't need to block on the sdsc finding that is only potentially relevant to symbolic shape support.  Can be addressed later.

**Comment by ani300** [comment]
@dgrove-oss size_hint doesn't exist anymore on PT 2.12, and it was split into both optimization_hint and guarding_hint_or_throw. I've pushed a fix that explicitly uses the throwing variant in the places you mentioned.

**Review by thoangtrvn — DISMISSED** [review]

**Review by dgrove-oss — DISMISSED** [review]
lgtm 🚀

File: torch_spyre/_inductor/__init__.py [hunk 5]
```diff
@@ -18,13 +18,35 @@
...
 
+def _spyre_inner_compile(*args: Any, **kwargs: Any) -> Any:
+    """Wrapper around ``compile_fx_inner`` that pins a picklable ``get_decomp_fn``.
+
+    Background: passing ``decompositions=<dict>`` to ``compile_fx`` causes it
+    to wrap the dict in a local ``def get_decomp_fn`` closure (compile_fx.py).
+    That closure is unpicklable, so the FX graph cache silently bypasses
+    itself with ``BypassFxGraphCache("Failed to pickle cache key")``.
+
+    Workaround: we never pass ``decompositions=``. Instead we override
```
Comment (JRosenkranz) [inline torch_spyre/_inductor/__init__.py]: Isn't decompositions being set here:

```python
kwargs.setdefault("decompositions", get_spyre_decomp_table())
```

This comment needs updating as it reads like it maybe a bug by setting it

File: torch_spyre/_inductor/patches.py [hunk 6]
```diff
@@ -37,40 +36,32 @@ def spyre_data_types():
...
     This CM configures PyTorch Inductor to compile graphs for the Spyre device by:
       - Enabling Spyre-specific data type handling
-      - Activating Spyre lowerings and decompositions
+      - Activating Spyre lowerings
       - Configuring Inductor settings optimized for Spyre
       - Setting up custom pre/post compilation passes
       - Disabling incompatible optimizations (e.g., reduction splitting, permute fusion)
 
+    Spyre-specific decompositions are *not* installed by this CM. They are
+    threaded into Inductor via ``get_decomp_fn`` (see ``_spyre_get_decomp_fn``
```
Comment (JRosenkranz) [inline torch_spyre/_inductor/patches.py]: Does `_spyre_get_decomp_fn` exist?

File: torch_spyre/_inductor/__init__.py (hunk 5 above)
Comment (ani300) [reply torch_spyre/_inductor/__init__.py]: fixed

File: torch_spyre/_inductor/patches.py (hunk 6 above)
Comment (ani300) [reply torch_spyre/_inductor/patches.py]: fixed

**Review by dgrove-oss — DISMISSED** [review]
lgtm

**Review by thoangtrvn — DISMISSED** [review]

**Review by anubhavjana — DISMISSED** [review]

**Review by JRosenkranz — DISMISSED** [review]
lgtm

**Review by mudhakar — DISMISSED** [review]

**Review by mudhakar — APPROVED** [review]

**Review by dgrove-oss — APPROVED** [review]

**Review by thoangtrvn — APPROVED** [review]

**Review by anubhavjana — APPROVED** [review]

**Review by JRosenkranz — APPROVED** [review]
Lgtm

**Review by pradghos — APPROVED** [review]
LGTM, Thanks!
